controllers/topics/index.js

```diff
@@ -102,6 +102,107 @@ const putFollowTopic = async (req, res) => {
   }
 };
 
+const followTopicV2Func = async (req) => {
+  const {name, with_system_message = false} = req.body;
+  const {token, userId} = req;
+
+  // logic get user sign and anonymous
+  let secondDetailUser;
+  let secondDetailUserId;
+  const detailTokenUser = await UsersFunction.findUserById(User, userId);
+  if (!detailTokenUser.is_anonymous) {
+    secondDetailUser = await UsersFunction.findAnonymousUserId(User, userId);
+  } else {
+    secondDetailUserId = await UsersFunction.findSignedUserId(User, userId);
+    secondDetailUser = await UsersFunction.findUserById(User, secondDetailUserId);
+  }
+
+  let prevUserToken = await createToken(secondDetailUser.user_id);
+  TopicValidator.validatePutTopicFollow({name});
+
+  //Logic get topic
+  const topicService = new TopicService(Topics);
+  const topic = await topicService.getTopicByName(name);
+  const {topic_id} = topic;
+  const userTopicService = new UserTopicService(UserTopic, UserTopicHistory);
+
+  //get follow status
+  const [getTokenUserStatus, getSecondUserStatus] = await Promise.all([
+    userTopicService.getFollowTopicStatus(detailTokenUser.user_id, topic_id),
+    userTopicService.getFollowTopicStatus(secondDetailUser.user_id, topic_id)
+  ]);
+
+  let topicInvitationIds = [];
+  let data = [];
+  if (getTokenUserStatus) {
+    await userTopicService.unfollowTopic(detailTokenUser.user_id, topic_id);
+    data.push(
+      await _afterPutTopic(
+        true,
+        token,
+        detailTokenUser.user_id,
+        name,
+        detailTokenUser.is_anonymous,
+        topicInvitationIds,
+        prevUserToken,
+        with_system_message
+      )
+    );
+  } else {
+    //get invitations topic detail
+    const topicInvitations = await TopicInvitations.findAll({
+      where: {
+        user_id_invited: userId,
+        topic_id
+      }
+    });
+
+    if (topicInvitations.length > 0) {
+      topicInvitationIds = topicInvitations.map(
+        (topicInvitation) => topicInvitation.topic_invitations_id
+      );
+      await TopicInvitations.destroy({
+        where: {
+          topic_invitations_id: topicInvitationIds
+        }
+      });
+    }
+
+    await userTopicService.followTopic(detailTokenUser.user_id, topic_id);
+    data.push(
+      await _afterPutTopic(
+        false,
+        token,
+        detailTokenUser.user_id,
+        name,
+        detailTokenUser.is_anonymous,
+        topicInvitationIds,
+        prevUserToken,
+        with_system_message
+      )
+    );
+
+    if (getSecondUserStatus) {
+      topicInvitationIds = [];
+      await userTopicService.unfollowTopic(secondDetailUser.user_id, topic_id);
+      data.push(
+        await _afterPutTopic(
+          true,
+          token,
+          secondDetailUser.user_id,
+          name,
+          detailTokenUser.is_anonymous,
+          topicInvitationIds,
+          prevUserToken,
+          with_system_message
+        )
+      );
+    }
+  }
+
+  return data;
+};
```

Comment:
<img src="https://avatars.githubusercontent.com/in/347564?s=41" alt="Image description" width="20" height="20">   CodeRabbit

In the `followTopicV2Func` function, there is a lack of error handling. If any of the promises fail, it will cause an unhandled promise rejection. It's recommended to wrap each async operation in a try-catch block or use Promise.allSettled instead of Promise.all to handle individual promise rejections.

```diff
- const [getTokenUserStatus, getSecondUserStatus] = await Promise.all([
-   userTopicService.getFollowTopicStatus(detailTokenUser.user_id, topic_id),
-   userTopicService.getFollowTopicStatus(secondDetailUser.user_id, topic_id)
- ]);
+ const [getTokenUserStatusResult, getSecondUserStatusResult] = await Promise.allSettled([
+   userTopicService.getFollowTopicStatus(detailTokenUser.user_id, topic_id),
+   userTopicService.getFollowTopicStatus(secondDetailUser.user_id, topic_id)
+ ]);

+ if (getTokenUserStatusResult.status === 'rejected') {
+   throw getTokenUserStatusResult.reason;
+ }
+ if (getSecondUserStatusResult.status === 'rejected') {
+   throw getSecondUserStatusResult.reason;
+ }

+ const getTokenUserStatus = getTokenUserStatusResult.value;
+ const getSecondUserStatus = getSecondUserStatusResult.value;
```



<!-- This is an auto-generated comment by OSS CodeRabbit -->